File: app/src/test/java/com/vrem/wifianalyzer/wifi/graphutils/GraphViewWrapperTest.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer.wifi.graphutils;

import android.graphics.Color;
import android.view.View;

import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GridLabelRenderer;
import com.jjoe64.graphview.LegendRenderer;
import com.jjoe64.graphview.Viewport;
import com.jjoe64.graphview.series.BaseSeries;
import com.jjoe64.graphview.series.DataPoint;
import com.vrem.wifianalyzer.Configuration;
import com.vrem.wifianalyzer.settings.ThemeStyle;
import com.vrem.wifianalyzer.wifi.model.WiFiDetail;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Collections;
import java.util.List;
import java.util.Set;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GraphViewWrapperTest {
    @Mock
    private GraphView graphView;
    @Mock
    private Viewport viewport;
    @Mock
    private GridLabelRenderer gridLabelRenderer;
    @Mock
    private LegendRenderer legendRenderer;
    @Mock
    private SeriesCache seriesCache;
    @Mock
    private SeriesOptions seriesOptions;
    @Mock
    private BaseSeries<DataPoint> baseSeries;

    private DataPoint dataPoint;
    private DataPoint[] dataPoints;
    private WiFiDetail wiFiDetail;
    private GraphViewWrapper fixture;

    @Before
    public void setUp() {
        wiFiDetail = WiFiDetail.EMPTY;
        dataPoint = new DataPoint(1, 2);
        dataPoints = new DataPoint[]{dataPoint};

        when(graphView.getLegendRenderer()).thenReturn(legendRenderer);

        fixture = new GraphViewWrapper(graphView, GraphLegend.HIDE, ThemeStyle.DARK) {
            @Override
            protected LegendRenderer newLegendRenderer() {
                return legendRenderer;
            }
        };
        fixture.setSeriesCache(seriesCache);
        fixture.setSeriesOptions(seriesOptions);

        assertEquals(GraphLegend.HIDE, fixture.getGraphLegend());
    }

    @Test
    public void testRemoveSeries() {
        // setup
        Set<WiFiDetail> newSeries = Collections.emptySet();
        List<WiFiDetail> difference = Collections.emptyList();
        List<BaseSeries<DataPoint>> removed = Collections.singletonList(baseSeries);
        when(seriesCache.difference(newSeries)).thenReturn(difference);
        when(seriesCache.remove(difference)).thenReturn(removed);
        // execute
        fixture.removeSeries(newSeries);
        // validate
        verify(seriesCache).difference(newSeries);
        verify(seriesCache).remove(difference);
        verify(seriesOptions).removeSeriesColor(baseSeries);
        verify(graphView).removeSeries(baseSeries);
    }

    @Test
    public void testDifferenceSeries() {
        // setup
        Set<WiFiDetail> newSeries = Collections.emptySet();
        List<WiFiDetail> expected = Collections.emptyList();
        when(seriesCache.difference(newSeries)).thenReturn(expected);
        // execute
        List<WiFiDetail> actual = fixture.differenceSeries(newSeries);
        // validate
        assertEquals(expected, actual);
        verify(seriesCache).difference(newSeries);
    }

    @Test
    public void testAddSeriesDirectly() {
        // execute
        fixture.addSeries(baseSeries);
        // validate
        verify(graphView).addSeries(baseSeries);
    }

    @Test
    public void testAddSeriesWhenSeriesExistsDoesNotAddSeries() {
        // setup
        when(seriesCache.contains(wiFiDetail)).thenReturn(true);
        // execute
        boolean actual = fixture.addSeries(wiFiDetail, baseSeries, false);
        // validate
        assertFalse(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache, never()).put(wiFiDetail, baseSeries);
    }

    @Test
    public void testAddSeriesAddsSeries() {
        // setup
        String expectedTitle = wiFiDetail.getSSID() + " " + wiFiDetail.getWiFiSignal().getChannelDisplay();
        boolean connected = wiFiDetail.getWiFiAdditional().getWiFiConnection().isConnected();
        when(seriesCache.contains(wiFiDetail)).thenReturn(false);
        // execute
        boolean actual = fixture.addSeries(wiFiDetail, baseSeries, true);
        // validate
        assertTrue(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache).put(wiFiDetail, baseSeries);
        verify(baseSeries).setTitle(expectedTitle);
        verify(baseSeries).setOnDataPointTapListener(any(GraphViewWrapper.GraphTapListener.class));
        verify(seriesOptions).highlightConnected(baseSeries, connected);
        verify(seriesOptions).setSeriesColor(baseSeries);
        verify(seriesOptions).drawBackground(baseSeries, true);
        verify(graphView).addSeries(baseSeries);
    }

    @Test
    public void testUpdateSeriesWhenSeriesDoesNotExistsDoesNotUpdateSeries() {
        // setup
        when(seriesCache.contains(wiFiDetail)).thenReturn(false);
        // execute
        boolean actual = fixture.updateSeries(wiFiDetail, dataPoints, true);
        // validate
        assertFalse(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache, never()).get(wiFiDetail);
    }

    @Test
    public void testUpdateSeriesWhenSeriesDoesExists() {
        // setup
        boolean connected = wiFiDetail.getWiFiAdditional().getWiFiConnection().isConnected();
        when(seriesCache.contains(wiFiDetail)).thenReturn(true);
        when(seriesCache.get(wiFiDetail)).thenReturn(baseSeries);
        // execute
        boolean actual = fixture.updateSeries(wiFiDetail, dataPoints, true);
        // validate
        assertTrue(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache).get(wiFiDetail);
        verify(baseSeries).resetData(dataPoints);
        verify(seriesOptions).highlightConnected(baseSeries, connected);
        verify(seriesOptions).drawBackground(baseSeries, true);
    }

    @Test
    public void testAppendSeriesWhenSeriesDoesNotExistsDoesNotUpdateSeries() {
        // setup
        int count = 10;
        when(seriesCache.contains(wiFiDetail)).thenReturn(false);
        // execute
        boolean actual = fixture.appendToSeries(wiFiDetail, dataPoint, count, true);
        // validate
        assertFalse(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache, never()).get(wiFiDetail);
    }

    @Test
    public void testAppendSeriesWhenSeriesDoesExists() {
        // setup
        int count = 10;
        boolean connected = wiFiDetail.getWiFiAdditional().getWiFiConnection().isConnected();
        when(seriesCache.contains(wiFiDetail)).thenReturn(true);
        when(seriesCache.get(wiFiDetail)).thenReturn(baseSeries);
        // execute
        boolean actual = fixture.appendToSeries(wiFiDetail, dataPoint, count, true);
        // validate
        assertTrue(actual);
        verify(seriesCache).contains(wiFiDetail);
        verify(seriesCache).get(wiFiDetail);
        verify(baseSeries).appendData(dataPoint, true, count + 1);
        verify(seriesOptions).highlightConnected(baseSeries, connected);
        verify(seriesOptions).drawBackground(baseSeries, true);
    }

    @Test
    public void testUpdateLegend() {
        // setup
        float textSize = 10f;
        when(graphView.getTitleTextSize()).thenReturn(textSize);
        // execute
        fixture.updateLegend(GraphLegend.RIGHT);
        // validate
        assertEquals(GraphLegend.RIGHT, fixture.getGraphLegend());
        verify(graphView).setLegendRenderer(legendRenderer);
        verify(legendRenderer).resetStyles();
        verify(legendRenderer).setWidth(0);
        verify(legendRenderer).setTextSize(textSize);
        verify(legendRenderer).setTextColor(Color.WHITE);
    }

    @Test
    public void testSetVisibility() {
        // execute
        fixture.setVisibility(View.VISIBLE);
        // validate
        verify(graphView).setVisibility(View.VISIBLE);
    }

    @Test
    public void testCalculateGraphType() {
        // execute & validate
        assertTrue(fixture.calculateGraphType() > 0);
    }

    @Test
    public void testSetViewport() {
        // setup
        when(graphView.getGridLabelRenderer()).thenReturn(gridLabelRenderer);
        when(gridLabelRenderer.getNumHorizontalLabels()).thenReturn(10);
        when(graphView.getViewport()).thenReturn(viewport);
        // execute
        fixture.setViewport();
        // validate
        verify(graphView).getGridLabelRenderer();
        verify(gridLabelRenderer).getNumHorizontalLabels();
        verify(graphView).getViewport();
        verify(viewport).setMinX(0);
        verify(viewport).setMaxX(9);
    }

    @Test
    public void testGetSize() {
        // execute & validate
        assertEquals(Configuration.SIZE_MAX, fixture.getSize(GraphConstants.TYPE1));
        assertEquals(Configuration.SIZE_MAX, fixture.getSize(GraphConstants.TYPE2));
        assertEquals(Configuration.SIZE_MAX, fixture.getSize(GraphConstants.TYPE3));
        assertEquals(Configuration.SIZE_MIN, fixture.getSize(GraphConstants.TYPE4));
    }

    @Test
    public void testSetViewportWithMinAndMax() {
        // setup
        when(graphView.getViewport()).thenReturn(viewport);
        // execute
        fixture.setViewport(1, 2);
        // validate
        verify(graphView).getViewport();
        verify(viewport).setMinX(1);
        verify(viewport).setMaxX(2);
    }

    @Test
    public void testIsNewSeries() {
        // setup
        when(seriesCache.contains(wiFiDetail)).thenReturn(false);
        // execute
        boolean actual = fixture.isNewSeries(wiFiDetail);
        // validate
        assertTrue(actual);
        verify(seriesCache).contains(wiFiDetail);
    }
}

File: app/src/test/kotlin/com/vrem/wifianalyzer/wifi/graphutils/GraphViewWrapperTest.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2026 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer.wifi.graphutils

import android.graphics.Color
import android.view.View
import com.jjoe64.graphview.GraphView
import com.jjoe64.graphview.GridLabelRenderer
import com.jjoe64.graphview.LegendRenderer
import com.jjoe64.graphview.Viewport
import com.jjoe64.graphview.series.BaseSeries
import com.vrem.wifianalyzer.SIZE_MAX
import com.vrem.wifianalyzer.SIZE_MIN
import com.vrem.wifianalyzer.settings.ThemeStyle
import com.vrem.wifianalyzer.wifi.model.WiFiDetail
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.spy
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever

class GraphViewWrapperTest {
    private val graphView: GraphView = mock()
    private val viewport: Viewport = mock()
    private val gridLabelRenderer: GridLabelRenderer = mock()
    private val legendRenderer: LegendRenderer = mock()
    private val seriesCache: SeriesCache = mock()
    private val seriesOptions: SeriesOptions = mock()
    private val baseSeries: BaseSeries<GraphDataPoint> = mock()
    private val dataPoint: GraphDataPoint = GraphDataPoint(1, 2)
    private val dataPoints = arrayOf(dataPoint)
    private val wiFiDetail = WiFiDetail.EMPTY
    private val fixture =
        spy(GraphViewWrapper(graphView, GraphLegend.HIDE, ThemeStyle.DARK, seriesCache, seriesOptions))

    @Before
    fun setUp() {
        assertThat(fixture.graphLegend).isEqualTo(GraphLegend.HIDE)
    }

    @After
    fun tearDown() {
        verifyNoMoreInteractions(viewport)
    }

    @Test
    fun removeSeries() {
        // setup
        val newSeries: Set<WiFiDetail> = setOf()
        val difference: List<WiFiDetail> = listOf()
        val removed = listOf(baseSeries)
        whenever(seriesCache.difference(newSeries)).thenReturn(difference)
        whenever(seriesCache.remove(difference)).thenReturn(removed)
        // execute
        fixture.removeSeries(newSeries)
        // validate
        verify(seriesCache).difference(newSeries)
        verify(seriesCache).remove(difference)
        verify(seriesOptions).removeSeriesColor(baseSeries)
        verify(graphView).removeSeries(baseSeries)
    }

    @Test
    fun differenceSeries() {
        // setup
        val newSeries: Set<WiFiDetail> = setOf()
        val expected: List<WiFiDetail> = listOf()
        whenever(seriesCache.difference(newSeries)).thenReturn(expected)
        // execute
        val actual = fixture.differenceSeries(newSeries)
        // validate
        assertThat(actual).isEqualTo(expected)
        verify(seriesCache).difference(newSeries)
    }

    @Test
    fun addSeriesDirectly() {
        // execute
        fixture.addSeries(baseSeries)
        // validate
        verify(graphView).addSeries(baseSeries)
    }

    @Test
    fun addSeriesWhenSeriesExistsDoesNotAddSeries() {
        // setup
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(true)
        // execute
        val actual = fixture.addSeries(wiFiDetail, baseSeries, false)
        // validate
        assertThat(actual).isFalse
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache, never()).put(wiFiDetail, baseSeries)
    }

    @Test
    fun addSeriesAddsSeries() {
        // setup
        val expectedTitle = wiFiDetail.wiFiIdentifier.ssid + " " + wiFiDetail.wiFiSignal.channelDisplay()
        val connected = wiFiDetail.wiFiAdditional.wiFiConnection.connected
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(false)
        // execute
        val actual = fixture.addSeries(wiFiDetail, baseSeries, true)
        // validate
        assertThat(actual).isTrue
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache).put(wiFiDetail, baseSeries)
        verify(baseSeries).title = expectedTitle
        verify(baseSeries).setOnDataPointTapListener(any())
        verify(seriesOptions).highlightConnected(baseSeries, connected)
        verify(seriesOptions).setSeriesColor(baseSeries)
        verify(seriesOptions).drawBackground(baseSeries, true)
        verify(graphView).addSeries(baseSeries)
    }

    @Test
    fun updateSeriesWhenSeriesDoesNotExistsDoesNotUpdateSeries() {
        // setup
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(false)
        // execute
        val actual = fixture.updateSeries(wiFiDetail, dataPoints, true)
        // validate
        assertThat(actual).isFalse
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache, never())[wiFiDetail]
    }

    @Test
    fun updateSeriesWhenSeriesDoesExists() {
        // setup
        val expectedTitle = wiFiDetail.wiFiIdentifier.ssid + " " + wiFiDetail.wiFiSignal.channelDisplay()
        val connected = wiFiDetail.wiFiAdditional.wiFiConnection.connected
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(true)
        whenever(seriesCache[wiFiDetail]).thenReturn(baseSeries)
        // execute
        val actual = fixture.updateSeries(wiFiDetail, dataPoints, true)
        // validate
        assertThat(actual).isTrue
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache)[wiFiDetail]
        verify(baseSeries).resetData(dataPoints)
        verify(baseSeries).title = expectedTitle
        verify(seriesOptions).highlightConnected(baseSeries, connected)
        verify(seriesOptions).drawBackground(baseSeries, true)
    }

    @Test
    fun appendSeriesWhenSeriesDoesNotExistsDoesNotUpdateSeries() {
        // setup
        val count = 10
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(false)
        // execute
        val actual = fixture.appendToSeries(wiFiDetail, dataPoint, count, true)
        // validate
        assertThat(actual).isFalse
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache, never())[wiFiDetail]
    }

    @Test
    fun appendSeriesWhenSeriesDoesExists() {
        // setup
        val count = 10
        val connected = wiFiDetail.wiFiAdditional.wiFiConnection.connected
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(true)
        whenever(seriesCache[wiFiDetail]).thenReturn(baseSeries)
        // execute
        val actual = fixture.appendToSeries(wiFiDetail, dataPoint, count, true)
        // validate
        assertThat(actual).isTrue
        verify(seriesCache).contains(wiFiDetail)
        verify(seriesCache)[wiFiDetail]
        verify(baseSeries).appendData(dataPoint, true, count + 1)
        verify(seriesOptions).highlightConnected(baseSeries, connected)
        verify(seriesOptions).drawBackground(baseSeries, true)
    }

    @Test
    fun updateLegend() {
        // setup
        val textSize = 10f
        doReturn(legendRenderer).whenever(fixture).newLegendRenderer()
        whenever(graphView.titleTextSize).thenReturn(textSize)
        whenever(graphView.legendRenderer).thenReturn(legendRenderer)
        // execute
        fixture.updateLegend(GraphLegend.RIGHT)
        // validate
        assertThat(fixture.graphLegend).isEqualTo(GraphLegend.RIGHT)
        verify(graphView).titleTextSize
        verify(graphView).legendRenderer
        verify(graphView).legendRenderer = legendRenderer
        verify(legendRenderer).resetStyles()
        verify(legendRenderer).width = 0
        verify(legendRenderer).textSize = textSize
        verify(legendRenderer).textColor = Color.WHITE
        verify(legendRenderer).isVisible = true
        verify(legendRenderer).align = LegendRenderer.LegendAlign.TOP
    }

    @Test
    fun setVisibility() {
        // execute
        fixture.visibility(View.VISIBLE)
        // validate
        verify(graphView).visibility = View.VISIBLE
    }

    @Test
    fun setHorizontalLabelsVisible() {
        // setup
        doReturn(gridLabelRenderer).whenever(graphView).gridLabelRenderer
        // execute
        fixture.setHorizontalLabelsVisible(true)
        // validate
        verify(graphView).gridLabelRenderer
        verify(gridLabelRenderer).isHorizontalLabelsVisible = true
    }

    @Test
    fun calculateGraphType() {
        // execute & validate
        assertThat(fixture.calculateGraphType()).isGreaterThan(0)
    }

    @Test
    fun setViewport() {
        // setup
        whenever(graphView.gridLabelRenderer).thenReturn(gridLabelRenderer)
        whenever(gridLabelRenderer.numHorizontalLabels).thenReturn(10)
        whenever(graphView.viewport).thenReturn(viewport)
        // execute
        fixture.setViewport()
        // validate
        verify(graphView).gridLabelRenderer
        verify(gridLabelRenderer).numHorizontalLabels
        verify(graphView).viewport
        verify(viewport).setMinX(0.0)
        verify(viewport).setMaxX(9.0)
    }

    @Test
    fun getSize() {
        // execute & validate
        assertThat(fixture.size(TYPE1)).isEqualTo(SIZE_MAX)
        assertThat(fixture.size(TYPE2)).isEqualTo(SIZE_MAX)
        assertThat(fixture.size(TYPE3)).isEqualTo(SIZE_MAX)
        assertThat(fixture.size(TYPE4)).isEqualTo(SIZE_MIN)
    }

    @Test
    fun setViewportSetsMinAndMaxX() {
        // setup
        whenever(graphView.viewport).thenReturn(viewport)
        whenever(graphView.gridLabelRenderer).thenReturn(gridLabelRenderer)
        whenever(gridLabelRenderer.numHorizontalLabels).thenReturn(5)
        // execute
        fixture.setViewport()
        // validate
        verify(viewport).setMinX(0.0)
        verify(viewport).setMaxX(4.0)
    }

    @Test
    fun setViewportWithParamsSetsMinAndMaxX() {
        // setup
        whenever(graphView.viewport).thenReturn(viewport)
        // execute
        fixture.setViewport(1, 10)
        // validate
        verify(viewport).setMinX(1.0)
        verify(viewport).setMaxX(10.0)
    }

    @Test
    fun viewportCntXReturnsCorrectValue() {
        // setup
        whenever(graphView.gridLabelRenderer).thenReturn(gridLabelRenderer)
        whenever(gridLabelRenderer.numHorizontalLabels).thenReturn(7)
        // execute
        assertThat(fixture.viewportCntX).isEqualTo(6)
        // validate
        verify(graphView).gridLabelRenderer
        verify(gridLabelRenderer).numHorizontalLabels
    }

    @Test
    fun updateLegendWhenSameLegendDoesNotResetLegendRenderer() {
        whenever(graphView.legendRenderer).thenReturn(legendRenderer)
        whenever(graphView.titleTextSize).thenReturn(12f)
        fixture.updateLegend(GraphLegend.HIDE)
        verify(legendRenderer).resetStyles()
    }

    @Test
    fun setHorizontalLabelsVisibleSetsValue() {
        // setup
        whenever(graphView.gridLabelRenderer).thenReturn(gridLabelRenderer)
        // execute
        fixture.setHorizontalLabelsVisible(true)
        // validate
        verify(gridLabelRenderer).isHorizontalLabelsVisible = true
    }

    @Test
    fun visibilitySetsVisibility() {
        // execute
        fixture.visibility(View.VISIBLE)
        // validate
        verify(graphView).visibility = View.VISIBLE
    }

    @Test
    fun newSeriesReturnsTrueIfNotExists() {
        // setup
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(false)
        // execute & validate
        assertThat(fixture.newSeries(wiFiDetail)).isTrue
    }

    @Test
    fun newSeriesReturnsFalseIfExists() {
        // setup
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(true)
        // execute & validate
        assertThat(fixture.newSeries(wiFiDetail)).isFalse
    }

    @Test
    fun addSeriesTriggersPopupOnTap() {
        // setup
        whenever(seriesCache.contains(wiFiDetail)).thenReturn(false)
        // execute
        fixture.addSeries(wiFiDetail, baseSeries, false)
        // validate
        verify(baseSeries).setOnDataPointTapListener(any())
    }
}
